### ISSUE TITLE
Branch history out of order after conversion

### ISSUE THREAD

**Issue by jnhunsberger** [frej/fast-export#71]
Hi. I just ran fast-export on several mercurial repos and am manually reviewing the commit and branch history. I noticed that everything seems in order on the default/master branch, but other branches have their commits out of order. This is confirmed by comparing both the graph history and the date sequences. It seems to be grouping some of the commits by author name and date ordering them. Repos that don't have branching seem to be fine. I have only run the command:

$ pwd
/path to git directory/
$ hg-fast-export.sh -r <hg repo> 

Is there another pass or command I need to run to clean-up the branch history?

**Comment by frej** [comment]
> Hi. I just ran fast-export on several mercurial repos and am manually
> reviewing the commit and branch history. I noticed that everything
> seems in order on the default/master branch, but other branches have
> their commits out of order. This is confirmed by comparing both the
> graph history and the date sequences. It seems to be grouping some of
> the commits by author name and date ordering them.

Can you give me a link to a hg repo that has this problem (or script
that builds one)?

> Is there another pass or command I need to run to clean-up the branch
> history?

That should not be necessary.

**Comment by jnhunsberger** [comment]
Thanks for your fast response. It will take me a bit to write-up a script that can recreate this. The more I look at it, however, what it looks like is that commits that happen later in time, but are next sequentially on a branch, sometimes get placed immediately after a branch is made, rather than when they occur in  time. As far as I can tell, the code is committing to the correct state, its just that the sequence is at times jumbled. This makes some of the commit comments or the graph flows seem out of place or different. I don't have reason to believe that the code ends up in an incorrect state, but it does make following the sequence of events harder/different in git than the record that exists in hg. 

Here are two screenshots showing the condition. The pre-migration shot shows the commit history and highlights in blue two commits 
Pre-migration (Hg)
![pre_migration_hg](https://cloud.githubusercontent.com/assets/12008959/16708645/354fdd24-45c9-11e6-8eae-1867a39497a9.png)

Post-migration (Git)
![post_migration_git](https://cloud.githubusercontent.com/assets/12008959/16708649/49550c68-45c9-11e6-87b8-62a392d96da4.png)

**Comment by frej** [comment]
> Thanks for your fast response. It will take me a bit to write-up a
> script that can recreate this.

Check if there is a mercurial equivalent to git fast-export --anonymize,
as that could save you some time.

> As far as I can tell, the code is committing to the correct state, its
> just that the sequence is at times jumbled. This makes some of the
> commit comments or the graph flows seem out of place or different. I
> don't have reason to believe that the code ends up in an incorrect
> state, but it does make following the sequence of events
> harder/different in git than the record that exists in hg.

If you take a look, using git log --pretty=raw, at the merge commits,
noting the order of the parents and then do the same on the hg
side. Does the order of parents differ? If the order is the same, the
graph is the same and what you are experiencing is a matter of how your
tool pretty-prints the graph.

**Comment by jnhunsberger** [comment]
Thanks for the tip on reviewing the log files at the terminal. It was enlightening. Here is a simple example of the difference I'm seeing pre- and post- migration. The screenshot on the left is the Hg log. The screenshot on the right is the Git log. I have annotated the commit sequence in each to show the differences in the flow. You can see the git log flips two of the commits prior to the merge. 

![side-by-side](https://cloud.githubusercontent.com/assets/12008959/16713415/b569bcd6-4674-11e6-948d-e65d1e043e67.png)

**Comment by frej** [comment]
Jason Hunsberger <notifications@github.com> writes:

> Thanks for the tip on reviewing the log files at the terminal. It was
> enlightening. Here is a simple example of the difference I'm seeing
> pre- and post- migration. The screenshot on the left is the Hg log.
> The screenshot on the right is the Git log. I have annotated the
> commit sequence in each to show the differences in the flow. You can
> see the git log flips two of the commits prior to the merge. 
>
> side-by-

The two commits which are merged in the commit map like this:

hg:13c0... <=> git:6744...
hg:0b8f.... <=> git:731f...

Checking the merge commit, both first and second parents match up,
i.e. the DAG is the same. So I'd say fast-export produces a correct
conversion.

Git log defaults to chronological ordering, but it looks as if hg uses
topological ordering. Run git log --topo-order and it will match what hg
prints out.

It's probably your graphical tool which displays the graphs
differently. Check if you can influence its ordering algorithm.

**Comment by frej** [comment]
Closing this as it appears to be  a GUI issue.

**Comment by jnhunsberger** [comment]
I ultimately needed to go in another direction with this project so I stopped pursuing this. I did find that the 'git log --topo-order' command still showed that same behavior as I was seeing in the GUI. That said, I did not find an instance where the ultimate state of the repository was somehow different. I just found some comments and submission dates that were no longer in 'order'. 

Anyway, thanks for your help. This is a great tool.
